Entities/TerrainDebris.cpp

```diff
@@ -84,104 +103,147 @@ namespace RTE {
+		writer.NewPropertyWithValue("MaxRotation", m_MaxRotation);
+		writer.NewPropertyWithValue("CanHFlip", m_CanHFlip);
+		writer.NewPropertyWithValue("CanVFlip", m_CanVFlip);
+		writer.NewPropertyWithValue("FlipChance", m_FlipChance);
+		writer.NewPropertyWithValue("DensityPerMeter", m_Density);
 
 		return 0;
 	}
 
 /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-	void TerrainDebris::ApplyDebris(SLTerrain *terrain) {
-		RTEAssert(!m_Bitmaps.empty() && m_BitmapCount > 0, "No bitmaps loaded for terrain debris!");
-
-		BITMAP *terrBitmap = terrain->GetFGColorBitmap();
+	bool TerrainDebris::GetPiecePlacementPosition(SLTerrain *terrain, Box &positionCheckBox) const {
 		BITMAP *matBitmap = terrain->GetMaterialBitmap();
-
-		// How many pieces of debris we're spreading out.
-		unsigned int terrainWidth = terrBitmap->w;
-		unsigned int pieceCount = static_cast<int>((static_cast<float>(terrainWidth) * c_MPP) * m_Density);
-
-		// First is index in the bitmap array, Second is blit location
-		std::list<std::pair<int, Vector>> piecesToPlace;
-		Vector location;
-		Box pieceBox;
-
-		unsigned char checkPixel;
-
-		acquire_bitmap(terrBitmap);
-		acquire_bitmap(matBitmap);
-
-		for (unsigned int piece = 0; piece < pieceCount; ++piece) {
-			bool place = false;
-			unsigned int currentBitmap = RandomNum<int>(0, m_BitmapCount - 1);
-			RTEAssert(currentBitmap >= 0 && currentBitmap < m_BitmapCount, "Bitmap index is out of bounds!");
-
-			pieceBox.SetWidth(static_cast<float>(m_Bitmaps.at(currentBitmap)->w));
-			pieceBox.SetHeight(static_cast<float>(m_Bitmaps.at(currentBitmap)->h));
-
-			int x = RandomNum<int>(0, terrainWidth);
-			int y = 0;
-			int depth = RandomNum(m_MinDepth, m_MaxDepth);
-
-			while (y < terrBitmap->h) {
-				// Find the air-terrain boundary
-				for (; y < terrBitmap->h; ++y) {
-					checkPixel = _getpixel(matBitmap, x, y);
-					// Check for terrain hit
-					if (checkPixel != g_MaterialAir) {
-						if (checkPixel == m_TargetMaterial.GetIndex()) {
-							place = true;
-							break;
-						// If we didn't hit target material, but are specified to, then don't place
-						} else if (m_OnlyOnSurface) {
-							place = false;
-							break;
+		int posX = RandomNum(0, matBitmap->w);
+		int depth = RandomNum(m_MinDepth, m_MaxDepth);
+		int buriedDepthOffset = m_OnlyBuried ? static_cast<int>(positionCheckBox.GetHeight() * 0.6F) : 0;
+		int prevMaterialCheckPixel = -1;
+
+		bool scanForOverhang = m_DebrisPlacementMode == DebrisPlacementMode::OnOverhangOnly || m_DebrisPlacementMode == DebrisPlacementMode::OnCavityOverhangOnly || m_DebrisPlacementMode == DebrisPlacementMode::OnOverhangAndCavityOverhang;
+
+		// For overhangs scan from the bottom so it's easier to detect.
+		for (int surfacePosY = 0, overhangPosY = matBitmap->h - 1; surfacePosY < matBitmap->h && overhangPosY > 0; surfacePosY++, overhangPosY--) {
```

Comment:
A prime example how actively avoiding code duplication leads to overcomplicated shit lel.
But it works and *mostly* straightforward.

Reply:
Yeah, no argument. It's not immediately obvious but given the concepts are moving in opposite directions, I think it's as straightforward as it can be without code duplication.
The other option would probably be to pull the code in the for loop into a function, and split the for loop up. But meh